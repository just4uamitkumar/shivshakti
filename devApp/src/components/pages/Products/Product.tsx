import { Grid, Stack } from "@mui/material";
import TypoGraphy from "../../common/TypoGraphy";

interface Props {
  id: number;
  title: string;
  thumbnail: string;
  price: number;
  discountPercentage: number;
  category: string;
  brand: string;
  shippingInformation: string;
  availabilityStatus: string;
}

const Product: React.FC<Props> = ({
  id,
  title,
  thumbnail,
  price,
  discountPercentage,
  category,
  brand,
  shippingInformation,
  availabilityStatus,
}) => {
  const getOriginalPrice = (
    discountPercent: number,
    discountedAmount: number
  ): number => {
    return discountedAmount / (1 - discountPercent / 100);
  };

  return (
    <>
      <Grid size={4} className="ProductCard" key={id}>
        <Stack className="imgWrapper">
          <img src={thumbnail} alt={title} />
        </Stack>
        <Stack>
          <TypoGraphy variant="h4">{title}</TypoGraphy>
        </Stack>
        <Stack className="price" direction={"row"} display={"flex"}>
          <Stack className="mr-2 buyPrice">
            <TypoGraphy variant="body1" typeClass="semi-bold-font">
              ${price}
            </TypoGraphy>
          </Stack>
          <Stack className="mr-2 line-through">
            <TypoGraphy variant="body1">
              {getOriginalPrice(discountPercentage, price).toFixed(2)}
            </TypoGraphy>
          </Stack>
          <Stack className="green500">
            <TypoGraphy variant="body1" typeClass="orange">
              {`${discountPercentage}% off`}
            </TypoGraphy>
          </Stack>
        </Stack>
        <Stack direction={"row"} display={"flex"}>
          <Stack className="mr-2" direction={"row"}>
            <TypoGraphy variant="body1">
              {"Category : "}
              {category}
            </TypoGraphy>
          </Stack>
          <Stack direction={"row"}>
            <TypoGraphy variant="body1">
              {"Brand : "}
              {brand}
            </TypoGraphy>
          </Stack>
        </Stack>
        <Stack className="shippingInfo">
          <TypoGraphy variant="body1">{shippingInformation}</TypoGraphy>
        </Stack>
        <Stack>
          <TypoGraphy variant="body1">
            <strong>{"Availability : "}</strong>
            {availabilityStatus}
          </TypoGraphy>
        </Stack>
      </Grid>
    </>
  );
};

export default Product;
