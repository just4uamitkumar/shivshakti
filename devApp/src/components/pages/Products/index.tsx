import { Grid, Stack } from "@mui/material";

import PageBanner from "../../shared/PageBanner";
import { useEffect, useState } from "react";
import axios from "axios";
import type { ProductType } from "./type";
import TypoGraphy from "../../common/TypoGraphy";
import {
  FaRegArrowAltCircleLeft,
  FaRegArrowAltCircleRight,
} from "react-icons/fa";

const Products: React.FC = () => {
  const [ProductsList, setProductsList] = useState<ProductType[]>([]);
  const [page, setPage] = useState(1);

  useEffect(() => {
    getProducts();
  }, []);

  const getProducts = async () => {
    const response = await axios.get(
      "https://dummyjson.com/products?limit=100"
    );
    console.log(response?.data?.products);
    setProductsList(response?.data?.products);
  };

  const selectPageHandler = (selectedPage: number) => {
    if (
      selectedPage >= 1 &&
      selectedPage <= ProductsList.length / 10 &&
      selectedPage !== page
    ) {
      setPage(selectedPage);
    }
  };

  return (
    <>
      <PageBanner title={"Products"} />

      <Grid className="container pb-4 pt-4" spacing={4} container>
        {ProductsList.length > 0 &&
          ProductsList.slice(page * 10 - 10, page * 10).map(
            (item: ProductType) => (
              <Grid size={6} className="ProductCard" key={item?.id}>
                <Stack className="imgWrapper">
                  <img src={item?.thumbnail} alt={item?.title} />
                </Stack>
                <Stack className="title">
                  <TypoGraphy variant="h4">{item?.title}</TypoGraphy>
                </Stack>
                <Stack className="price" direction={"row"} display={"flex"}>
                  <strong>{"Price : "}</strong>
                  <TypoGraphy variant="body1">{item?.price}</TypoGraphy>
                </Stack>
                <Stack className="price" direction={"row"} display={"flex"}>
                  <strong>{"Category : "}</strong>
                  <TypoGraphy variant="body1">{item?.category}</TypoGraphy>
                </Stack>
                <Stack className="Description">
                  <TypoGraphy variant="body1">{item?.description}</TypoGraphy>
                </Stack>
              </Grid>
            )
          )}
        <Grid size={12}>
          {ProductsList.length > 0 && (
            <Stack
              className="pagination"
              direction={"row"}
              display={"flex"}
              justifyContent={"center"}
            >
              <span
                onClick={() => selectPageHandler(page - 1)}
                className={page > 1 ? "" : "disable"}
              >
                <FaRegArrowAltCircleLeft />
              </span>

              {[...Array(ProductsList.length / 10)].map((_, i) => {
                return (
                  <span
                    key={i}
                    className={page === i + 1 ? "activePage" : ""}
                    onClick={() => selectPageHandler(i + 1)}
                  >
                    {i + 1}
                  </span>
                );
              })}

              <span
                onClick={() => selectPageHandler(page + 1)}
                className={
                  page < ProductsList.length / 10 ? "" : "disable"
                }
              >
                <FaRegArrowAltCircleRight />
              </span>
            </Stack>
          )}
        </Grid>
      </Grid>
    </>
  );
};

export default Products;
