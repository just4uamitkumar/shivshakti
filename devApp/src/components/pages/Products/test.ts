const myText = document.getElementById('myText');

function showMessage(event) {
     const myVal = event.target.value;
      console.log(myVal)
      myText.innerHTML = myVal
}
let typingTimer;

function myFunction(event){
  clearTimeout(typingTimer);
  typingTimer = setTimeout(function() {
      showMessage(event)
      }, 5000);
  
}

<input type="text" id="myInput" value='' onkeyup="myFunction(event)"/>
<div id="myText">Amit</div>