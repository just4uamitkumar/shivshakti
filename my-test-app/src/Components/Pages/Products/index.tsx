import { useEffect, useState } from "react";

type Product = {
  id: number;
  title: string;
  description: string;
  price: number;
  category: string;
  brand: string;
};

const Products = () => {
  const [productsList, setProductsList] = useState<Product[]>([]);
  const [loading, setLoading] = useState(false);

  useEffect(() => {
    getProducts();
  }, []);

  const getProducts = async () => {
    setLoading(true);
    const response = await fetch("https://dummyjson.com/products?limit=100");
    const data = await response.json();
    setProductsList(data?.products);
    setLoading(false);
  };

  console.log(productsList);

  return (
    <div className="products">
      <div className="container">
        <h1>Products</h1>
        {loading ? (
          <p>Loading...</p>
        ) : (
          <table className="products-table">
            <thead>
              <tr>
                <th>Sr. No.</th>
                <th>Name</th>
                <th>Description</th>
                <th>Price</th>
                <th>Category</th>
                <th>Brand</th>
              </tr>
            </thead>
            <tbody>
              {productsList.length > 0 &&
                productsList.map((product, index) => (
                  <tr key={product.id}>
                    <td>{index + 1}</td>
                    <td>{product.title}</td>
                    <td>{product.description}</td>
                    <td>${product.price}</td>
                    <td>{product.category}</td>
                    <td>{product.brand}</td>
                  </tr>
                ))}
            </tbody>
          </table>
        )}
      </div>
    </div>
  );
};

export default Products;
