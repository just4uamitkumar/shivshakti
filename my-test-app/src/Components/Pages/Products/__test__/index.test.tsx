import { describe, it, expect, beforeEach, afterEach, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import Products from "../index";

// Mock fetch or axios if your Products component fetches data
beforeEach(() => {
  globalThis.fetch = vi.fn(() =>
    Promise.resolve({
      json: () =>
        Promise.resolve({
          products: [
            {
              id: 1,
              title: "Product A",
              description: "Description A",
              price: 100,
              category: "Category A",
              brand: "Brand A",
            },
            {
              id: 2,
              title: "Product B",
              description: "Description B",
              price: 200,
              category: "Category B",
              brand: "Brand B",
            },
          ],
        }),
    })
  ) as unknown as typeof fetch;
});

afterEach(() => {
  vi.resetAllMocks();
});

describe("Products Component", () => {
  it("matches snapshot", () => {
    const { container } = render(<Products />);
    expect(container).toMatchSnapshot();
  });

  it("should render without crashing", () => {
    render(<Products />);
    expect(Products).toBeDefined();
  });

  it("should display the headline", () => {
    render(<Products />);
    const heading = screen.getByText("Products");
    expect(heading).toBeInTheDocument();
  });

  it("should render products from API", async () => {
    render(<Products />);
    expect(await screen.findByText("Product A")).toBeInTheDocument();
    expect(await screen.findByText("Product B")).toBeInTheDocument();
  });
});
