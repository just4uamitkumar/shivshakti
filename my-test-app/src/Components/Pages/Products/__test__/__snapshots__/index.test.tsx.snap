// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Products Component > matches snapshot 1`] = `
<div>
  <div
    class="products"
  >
    <div
      class="container"
    >
      <h1>
        Products
      </h1>
      <p>
        Loading...
      </p>
    </div>
  </div>
</div>
`;
